compiler/rustc_symbol_mangling/src/typeid.rs

```diff
@@ -67,8 +67,13 @@ pub fn kcfi_typeid_for_fnabi<'tcx>(
 pub fn kcfi_typeid_for_instance<'tcx>(
     tcx: TyCtxt<'tcx>,
     instance: Instance<'tcx>,
-    options: TypeIdOptions,
+    mut options: TypeIdOptions,
 ) -> u32 {
+    // If we receive a `ReifyShim` intended to produce a function pointer, we need to remain
+    // concrete - abstraction is for vtables.
+    if matches!(instance.def, InstanceDef::ReifyShim(_, Some(ReifyReason::FnPtr))) {
+        options |= TypeIdOptions::NO_SELF_TYPE_ERASURE
```

Comment:
it feels weird to have a negative option. Totally random ask, but is there a way to flip the polarity here? Maybe have a `default()` on `TypeIdOptions` with `SELF_TYPE_ERASURE` set and then remove it here?

Reply:
Shouldn't be too hard, I'll do it in a precursor commit.